epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java

```diff
@@ -165,24 +158,33 @@ private boolean checkAndUpdateFocusedVisible() {
    * not occupy at least half the viewport.
    */
   private boolean checkAndUpdateUnfocusedVisible(boolean detachEvent) {
-    boolean unfocusedVisible = detachEvent
-        || !(sizeInScrollingDirection >= viewportSize / 2 || (
-        visibleSize == sizeInScrollingDirection && sizeInScrollingDirection < viewportSize / 2));
+    boolean unfocusedVisible = detachEvent || !isInFocusVisible();
     if (unfocusedVisible) {
       focusedVisible = false;
     }
     return !focusedVisible;
   }
 
+  private boolean isInFocusVisible() {
+    final int halfViewportArea = viewportHeight * viewportWidth / 2;
+    final int totalArea = height * width;
+    final int visibleArea = visibleHeight * visibleWidth;
+    // The model has entered the focused range either if it is larger than half of the viewport
+    // and it occupies at least half of the viewport or if it is smaller than half of the viewport
+    // and it is fully visible.
+    return focusedVisible = (totalArea >= halfViewportArea)
```

Comment:
it is pretty confusing and fragile that `focusedVisible` is updated here, despite also being updated from the places that call this. Tracing the side effects and intent is difficult.

would it be possible to replace all of the methods to update and check focus with an approach that only mutates `focusedVisible` in one place?

```
    void handleFocus(EpoxyViewHolder epoxyHolder, boolean detachEvent) {
        boolean previousFocusedVisible = focusedVisible;
        focusedVisible = !detachEvent && isInFocusVisible();

        if (focusedVisible != previousFocusedVisible) {
            if (focusedVisible) {
                epoxyHolder.visibilityStateChanged(VisibilityState.FOCUSED_VISIBLE);
            } else {
                epoxyHolder.visibilityStateChanged(VisibilityState.UNFOCUSED_VISIBLE);
            }
        }
    }

    private boolean isInFocusVisible() {
        final int halfViewportArea = viewportHeight * viewportWidth / 2;
        final int totalArea = height * width;
        final int visibleArea = visibleHeight * visibleWidth;
        // The model has entered the focused range either if it is larger than half of the viewport
        // and it occupies at least half of the viewport or if it is smaller than half of the viewport
        // and it is fully visible.
        return (totalArea >= halfViewportArea)
            ? (visibleArea >= halfViewportArea)
            : totalArea == visibleArea;
    }
```

Reply:
This was a mistake, thanks for spotting it and for the code suggestion!